server/src/main/java/org/opensearch/identity/MyShiroModule.java

```diff
@@ -0,0 +1,118 @@
+import org.apache.shiro.subject.SimplePrincipalCollection;
+import org.apache.shiro.subject.Subject;
+import org.opensearch.common.Strings;
+
+import java.util.Base64;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Optional;
+
+/**
+ * Prototyping use of identity management (authentication & authorization) implemented with Shiro
+ * See more at https://shiro.apache.org/
+ */
+public class MyShiroModule {
+
+    public MyShiroModule() {
+        final Realm realm = new MyRealm();
+        final SecurityManager securityManager = new DefaultSecurityManager(realm);
+        // Note; this sets up security for the JVM, if we are crossing a JVM boundary will need to look at how this is made available
+        SecurityUtils.setSecurityManager(securityManager);
+    }
+
+    /**
+     * Generates an internal subject to run requests in this context
+     */
+    public static Subject getSubjectOrInternal() {
+        final Subject existingSubject = SecurityUtils.getSubject();
+        if (existingSubject.isAuthenticated()) {
+            return existingSubject;
+        }
+
+        // Used for tracing where this function is used that is visible in the log output
+        final Exception e = new Exception();
+        final StackTraceElement current = e.getStackTrace()[1];
+        final String sourceAnnotation = current.getFileName() + "." + current.getMethodName() + "@" + current.getLineNumber();
+
+        final Subject internalSubject = new Subject.Builder().authenticated(true)
+            .principals(new SimplePrincipalCollection("INTERNAL-" + sourceAnnotation, "OpenSearch")) // How can we ensure the roles this
+                                                                                                     // princpal resolves?
+            .contextAttribute("NodeId", "???") // Can we use this to source the originating node in a cluster?
+            .buildSubject();
+        return internalSubject;
+    }
+
+    /**
+     * Attempt to authenticate via authorization header, ignores if already authenticated, throws exceptions if unable
+     *
+     * NOTE: this was quickly built for test scenarios and will not be used log term, there is a supplimental library for
+     * Shiro that supports web based applications, including https://shiro.apache.org/static/1.9.1/apidocs/org/apache/shiro/web/filter/authc/BasicHttpAuthenticationFilter.html
+     * if this can be reused and to what degree will be useful in future investigations.
+     */
+    public static void authenticateViaAuthorizationHeader(final Optional<String> authHeader) throws AuthenticationException {
+        final Subject currentSubject = SecurityUtils.getSubject();
+        if (currentSubject.isAuthenticated()) {
+            // No need to authenticate already ready
+            return;
+        }
+
+        if (authHeader.isPresent() && !(Strings.isNullOrEmpty(authHeader.get())) && authHeader.get().startsWith("Basic ")) {
```

Comment:
I'm curious why the `authHeader.isPresent() || Strings.isNullOrEmpty(authHeader.get())` check isn't first in this method. Does every request have the Authorization header or do some requests not contain the header?

Reply:
Lets not look too deep into this code - its only purpose is to provide authentication in a way where we can see all the bits and pieces.

I am already ashamed of it